### DISCUSSION TITLE
Caddy

### DISCUSSION THREAD

**Discussion by BrunoTeixeira1996** [gokrazy/gokrazy#281] [Q&A] [answered]
Hi everyone,
So I have a few services running on gokrazy on different ports and I am having trouble remembering all ports so I wanted a way to map those ports to names.

Lets say I have the following ports - 8080 (web server), 8181 (web server 2), 8282 (web server 3). Currently I navigate to brun0-pi:8080 and so on.

But my idea was to do something like webserver.brun0-pi and go to port 8080.

I know Caddy is able to do this, however I tried a bunch of stuff and still did not worked.
Does anyone do the exact same thing and if yes, can you guys provide an example Cady file configuration so I can replicate on my gokrazy instance?

Thank you!

**Comment by stapelberg** [comment]
I think you’re asking about **reverse proxying**?

I use the following Caddyfile on one of my gokrazy instances to serve an application that listens on port 8319 on port 80 on a public address:

```
http://example.zekjur.net:80 {
	handle /* {
		reverse_proxy /* http://localhost:8319
	}
}
```

(Of course, listening on port 80 means you need to change gokrazy’s HTTP port as described in https://gokrazy.org/packages/caddy-http-server/)

**Reply by BrunoTeixeira1996** [reply]
Yes a reverse proxy!

But I have a few local apps running on gokrazy on different ports
Lets say I have port 8080 and port 8181 that run a web server each.
Can I use something like

```
http://webserver1.brun0.lan:80 {
	handle /* {
		reverse_proxy /* http://localhost:8080
	}
}

http://webserver2.brun0.lan:80 {
	handle /* {
		reverse_proxy /* http://localhost:8181
	}
}
```

And use `webserver1.brun0.lan` on a browser? I think I tried that on a VM and for some reason did not worked

**Reply by stapelberg** [reply]
That should work (technical term “virtual hosts”), provided `webserver1.brun0.lan` and `webserver2.brun0.lan` are both resolvable host names that point to the same IP address. Adding them to /etc/hosts on the machine on which you try to open them should work, at least for a quick test.

**Reply by BrunoTeixeira1996** [reply]
I am testing this and in fact works for stuff inside gokrazy (localhost) however I have several app outside gokrazy and for some reason that does not work. It simply does not resolv.
@stapelberg do you have any similar use case to mine?

**Reply by stapelberg** [reply]
I have successfully used Caddy to reverse proxy to other programs, some even running on other machines.

> I am testing this and in fact works for stuff inside gokrazy (localhost) however I have several app outside gokrazy and for some reason that does not work. It simply does not resolv.

Can you provide a lot more details? What are the steps to reproduce? What are the exact error messages? What did you try?

**Reply by BrunoTeixeira1996** [reply]
Sure I can!

so I am using the following Caddyfile inside gokrazy

````

http://gmah.brun0.lan:80 {
	handle /* {
		reverse_proxy /* http://localhost:9090
	}
}

http://cam.brun0.lan:80 {
	handle /* {
		reverse_proxy /* http://192.168.30.23:9999
	}
}
````

Then I have a LXC container in Proxmox that is running `dnsmasq` and in there I configured the local admin `brun0.lan` so this LXC is my LAN DNS Server.

On my laptop I can reach http://gmah.brun0.lan by using a browser and by making sure `dnsmasq` is mapping that subdomain to my Gokrazy IP.
That works totaly fine, however the `http://cam.brun0.lan` does not work. I simply get ERR_NO_RESPONSE and I can see that I can reach `http://cam.brun0.lan:9999` meaning that dnsmasq is working as intended.

I think the problem relies on Caddy but i am not 100% that I know this is out of scope for Gokrazy but maybe you can help in troubleshoot this.
In terms of error in caddy logs I have none.
So basicaly I can redirect every service running inside gokrazy but I cannot to the same to other servers outside the gokrazy environment

> I have successfully used Caddy to reverse proxy to other programs, some even running on other machines.

Hmmm I am wondering how Caddy receives that request to other machines, because if I try to go to http://cam.brun0.lan i think DNS will resolve directly to 192.168.30.23 and skip Caddy @stapelberg

**Reply by stapelberg** [reply]
> On my laptop I can reach http://gmah.brun0.lan by using a browser and by making sure `dnsmasq` is mapping that subdomain to my Gokrazy IP. That works totaly fine, however the `http://cam.brun0.lan` does not work. I simply get ERR_NO_RESPONSE and I can see that I can reach `http://cam.brun0.lan:9999` meaning that dnsmasq is working as intended.
> 
> I think the problem relies on Caddy but i am not 100% that I know this is out of scope for Gokrazy but maybe you can help in troubleshoot this. In terms of error in caddy logs I have none. So basicaly I can redirect every service running inside gokrazy but I cannot to the same to other servers outside the gokrazy environment
>
> Hmmm I am wondering how Caddy receives that request to other machines, because if I try to go to http://cam.brun0.lan i think DNS will resolve directly to 192.168.30.23 and skip Caddy @stapelberg

On your PC, what is the result of `host gmah.brun0.lan` and `host cam.brun0.lan`?

It sounds like cam.brun0.lan is not actually resolving to your gokrazy IP address. That is a requirement for requests to end up at Caddy to begin with, as you wonder.

> In terms of error in caddy logs I have none.

Any logs will be helpful. When Caddy starts up, it prints a few lines like these (including the listener configuration):

> 1970/01/01 01:00:03 gokrazy: attempt 0, starting ["/user/caddy" "run" "--config" "/etc/caddy/Caddyfile"]
> {"level":"info","ts":6.044450181,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":""}
> {"level":"info","ts":6.051342052,"logger":"admin","msg":"admin endpoint started","address":"localhost:2019","enforce_origin":false,"origins":["//localhost:2019","//[::1]:2019","//127.0.0.1:2019"]}
> {"level":"info","ts":6.052286867,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0x4000497380"}

**Reply by BrunoTeixeira1996** [reply, marked as answer]
The problem was infact on the dnsmasq side. I add to point `cam.brun0.lan` to Caddy!
